<launch>
    
    <param name="scan_line" type="int" value="64" />

    <!-- if 1, do mapping 10 Hz, if 2, do mapping 5 Hz. Suggest to use 1, it will adjust frequence automaticlly -->
    <param name="mapping_skip_frame" type="int" value="1" />

    <!-- remove too closed points -->
    <param name="minimum_range" type="double" value="5"/>

    <param name="mapping_line_resolution" type="double" value="0.1"/> <!-- A-LOAM -->
    <param name="mapping_plane_resolution" type="double" value="0.2"/> <!-- A-LOAM -->

    <!-- keypoint extraction -->
    <param name="LocalMapBoundary" type="double" value="25.0"/>
    <param name="KeypointsExtractionSkipFrame" type="int" value="1"/>

    <param name="HarrisRadius" type="double" value="1.0"/>
    <param name="HarrisRadiusSearch" type="double" value="30"/>
    <param name="HarrisThreshold" type="double" value="1e-8"/>

    <!-- SC-A-LOAM -->
    <param name="keyframe_meter_gap" type="double" value="3.0"/> 

    <param name="SaveMode" type="bool" value="true"/>

    <!-- Scan Context -->
    <param name="sc_dist_thres" type="double" value="0.2"/> <!-- SC-A-LOAM, if want no outliers, use 0.1-0.15 -->
    <!-- <param name="sc_max_radius" type="double" value="20.0"/> 20 or 40 for indoor -->
    <param name="sc_max_radius" type="double" value="40.0"/> <!-- for outdoor -->
    <param name="sc_source" type="string" value="keyframe"/> <!-- keyframe or denseFrame -->
    
    <!-- utils -->
    <param name="save_directory" type="string" value="/home/user/Documents/scaloam_scd_saver/data/00/"/>  <!-- CHANGE THIS and end with / -->

    <!--  -->
    <param name="lidar_type" type="string" value="HDL64"/> 
    <remap from="/velodyne_points" to="/velodyne_points"/>

    <node pkg="aloam_velodyne" type="ascanRegistration" name="ascanRegistration" output="screen" />
    <node pkg="aloam_velodyne" type="alaserOdometry" name="alaserOdometry" output="screen" />
    <node pkg="aloam_velodyne" type="alaserMapping" name="alaserMapping" output="screen" />
    <node pkg="aloam_velodyne" type="alaserPGO" name="alaserPGO" output="screen" />
    <node pkg="aloam_velodyne" type="alaserSC" name="alaseraSC" output="screen" />
    <!-- <node pkg="aloam_velodyne" type="alaserScDetector" name="alaserScDetector" output="screen" /> -->

    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find aloam_velodyne)/rviz_cfg/aloam_velodyne.rviz" />
    </group>

</launch>
